File: src/validate.zig
const std = @import("std");

const max_id_len = 64;
const max_version_len = 64;

/// Tool IDs: lowercase alphanumeric, hyphens, underscores only. Max 64 chars.
pub fn isValidToolId(id: []const u8) bool {
    if (id.len == 0 or id.len > max_id_len) return false;
    for (id) |c| {
        if (!std.ascii.isAlphanumeric(c) and c != '-' and c != '_') return false;
    }
    return true;
}

/// Version strings: alphanumeric, dots, hyphens, plus. No path separators or spaces.
pub fn isValidVersion(v: []const u8) bool {
    if (v.len == 0 or v.len > max_version_len) return false;
    for (v) |c| {
        if (!std.ascii.isAlphanumeric(c) and c != '.' and c != '-' and c != '+') return false;
    }
    return true;
}

// ─── Tests ────────────────────────────────────────────────────────────────────

test "isValidToolId: valid IDs" {
    try std.testing.expect(isValidToolId("helm"));
    try std.testing.expect(isValidToolId("kubectl"));
    try std.testing.expect(isValidToolId("my-tool"));
    try std.testing.expect(isValidToolId("my_tool"));
    try std.testing.expect(isValidToolId("tool123"));
    try std.testing.expect(isValidToolId("k9s"));
}

test "isValidToolId: invalid IDs" {
    try std.testing.expect(!isValidToolId(""));
    try std.testing.expect(!isValidToolId("tool with spaces"));
    try std.testing.expect(!isValidToolId("tool;rm -rf /"));
    try std.testing.expect(!isValidToolId("../evil"));
    try std.testing.expect(!isValidToolId("tool/path"));
    try std.testing.expect(!isValidToolId("tool\x00null"));
    try std.testing.expect(!isValidToolId("$(evil)"));
    // 65 chars — one over the limit
    try std.testing.expect(!isValidToolId("a" ** 65));
}

test "isValidVersion: valid versions" {
    try std.testing.expect(isValidVersion("3.15.0"));
    try std.testing.expect(isValidVersion("latest"));
    try std.testing.expect(isValidVersion("1.0.0-rc.1"));
    try std.testing.expect(isValidVersion("1.0.0+build.1"));
    try std.testing.expect(isValidVersion("v3.15.0"));
    try std.testing.expect(isValidVersion("2024.01.15"));
}

test "isValidVersion: invalid versions" {
    try std.testing.expect(!isValidVersion(""));
    try std.testing.expect(!isValidVersion("3.0/../../evil"));
    try std.testing.expect(!isValidVersion("1.0 && rm -rf /"));
    try std.testing.expect(!isValidVersion("$(evil)"));
    try std.testing.expect(!isValidVersion("1.0;evil"));
    try std.testing.expect(!isValidVersion("1.0\nnewline"));
    try std.testing.expect(!isValidVersion("a" ** 65));
}

File: src/validate.rs
const MAX_ID_LEN: usize = 64;
const MAX_VERSION_LEN: usize = 64;

pub fn is_valid_tool_id(id: &str) -> bool {
    if id.is_empty() || id.len() > MAX_ID_LEN {
        return false;
    }
    id.chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_')
}

pub fn is_valid_version(v: &str) -> bool {
    if v.is_empty() || v.len() > MAX_VERSION_LEN {
        return false;
    }
    v.chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '.' || c == '-' || c == '+')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn valid_tool_ids() {
        assert!(is_valid_tool_id("helm"));
        assert!(is_valid_tool_id("kubectl"));
        assert!(is_valid_tool_id("my-tool"));
        assert!(is_valid_tool_id("my_tool"));
        assert!(is_valid_tool_id("tool123"));
        assert!(is_valid_tool_id("k9s"));
    }

    #[test]
    fn invalid_tool_ids() {
        assert!(!is_valid_tool_id(""));
        assert!(!is_valid_tool_id("tool with spaces"));
        assert!(!is_valid_tool_id("tool;rm -rf /"));
        assert!(!is_valid_tool_id("../evil"));
        assert!(!is_valid_tool_id("tool/path"));
        assert!(!is_valid_tool_id("tool\x00null"));
        assert!(!is_valid_tool_id("$(evil)"));
        assert!(!is_valid_tool_id(&"a".repeat(65)));
    }

    #[test]
    fn valid_versions() {
        assert!(is_valid_version("3.15.0"));
        assert!(is_valid_version("latest"));
        assert!(is_valid_version("1.0.0-rc.1"));
        assert!(is_valid_version("1.0.0+build.1"));
        assert!(is_valid_version("v3.15.0"));
        assert!(is_valid_version("2024.01.15"));
    }

    #[test]
    fn invalid_versions() {
        assert!(!is_valid_version(""));
        assert!(!is_valid_version("3.0/../../evil"));
        assert!(!is_valid_version("1.0 && rm -rf /"));
        assert!(!is_valid_version("$(evil)"));
        assert!(!is_valid_version("1.0;evil"));
        assert!(!is_valid_version("1.0\nnewline"));
        assert!(!is_valid_version(&"a".repeat(65)));
    }
}
